src/components/calendar/types.rs

```diff
@@ -0,0 +1,397 @@
+            MONTH_TWELVE_LEAP,
+        ];
+
+        if COMMON_MONTH_CODES.contains(month_code) {
+            return Ok(MonthCode(*month_code));
+        }
+
+        match calendar.identifier() {
+            "chinese" | "dangi" if LUNAR_LEAP_MONTHS.contains(month_code) => {
+                Ok(MonthCode(*month_code))
+            }
+            "coptic" | "ethiopic" | "ethiopicaa" if MONTH_THIRTEEN == *month_code => {
+                Ok(MonthCode(*month_code))
+            }
+            "hebrew" if MONTH_FIVE_LEAP == *month_code => Ok(MonthCode(*month_code)),
+            _ => Err(TemporalError::range()
+                .with_message("MonthCode was not valid for the current calendar.")),
+        }
+    }
+
+    pub(crate) fn try_from_partial_date(
+        partial_date: &PartialDate,
+        calendar: &Calendar,
+    ) -> TemporalResult<Self> {
+        match partial_date {
+            PartialDate {
+                month: Some(month),
+                month_code: None,
+                ..
+            } => Self::try_new(&month_to_month_code(*month)?, calendar),
+            PartialDate {
+                month_code: Some(month_code),
+                month: None,
+                ..
+            } => Self::try_new(month_code, calendar),
+            PartialDate {
+                month: Some(month),
+                month_code: Some(month_code),
+                ..
+            } => {
+                are_month_and_month_code_resolvable(*month, month_code)?;
+                Self::try_new(month_code, calendar)
+            }
+            _ => Err(TemporalError::r#type()
+                .with_message("Month or monthCode is required to determine date.")),
+        }
+    }
+
+    pub fn as_iso_month_integer(&self) -> TemporalResult<u8> {
+        ascii_four_to_integer(self.0)
+    }
+}
+
+// NOTE: This is a greedy function, should handle differently for all calendars.
+pub(crate) fn month_to_month_code(month: i32) -> TemporalResult<TinyAsciiStr<4>> {
+    match month {
+        1 => Ok(MONTH_ONE),
+        2 => Ok(MONTH_TWO),
+        3 => Ok(MONTH_THREE),
+        4 => Ok(MONTH_FOUR),
+        5 => Ok(MONTH_FIVE),
+        6 => Ok(MONTH_SIX),
+        7 => Ok(MONTH_SEVEN),
+        8 => Ok(MONTH_EIGHT),
+        9 => Ok(MONTH_NINE),
+        10 => Ok(MONTH_TEN),
+        11 => Ok(MONTH_ELEVEN),
+        12 => Ok(MONTH_TWELVE),
+        13 => Ok(MONTH_THIRTEEN),
+        _ => Err(TemporalError::range().with_message("Month not in a valid range.")),
+    }
+}
+
+fn are_month_and_month_code_resolvable(month: i32, mc: &TinyAsciiStr<4>) -> TemporalResult<()> {
+    if month != ascii_four_to_integer(*mc)?.into() {
+        return Err(TemporalError::range()
+            .with_message("Month and monthCode values could not be resolved."));
+    }
+    Ok(())
+}
+
+// NOTE: This is a greedy function, should handle differently for all calendars.
+pub(crate) fn ascii_four_to_integer(mc: TinyAsciiStr<4>) -> TemporalResult<u8> {
```

Comment:
It still looks like this could return the wrong value on lunisolar calendars?

Reply:
Yeah 😕 it's a rather greedy function that I'd like to come back to in the future. It should be isolated to being called from solely ISO functions so it should be fine.